core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/NodeBuildHelper.java

```diff
@@ -74,78 +67,52 @@ public void addNode(AbstractNode node) {
         reversedNodeList.add(node);
     }
 
-    public <Tuple_ extends Tuple> void putInsertUpdateRetract(ConstraintStream stream, Consumer<Tuple_> insert,
-            Consumer<Tuple_> update, Consumer<Tuple_> retract) {
-        insertMap.put(stream, insert);
-        updateMap.put(stream, update);
-        retractMap.put(stream, retract);
+    public void addNode(AbstractNode node, ConstraintStream parent) {
+        addNode(node);
+        putInsertUpdateRetract(parent, (TupleLifecycle<? extends Tuple>) node);
     }
 
-    public <Tuple_ extends Tuple> void putInsertUpdateRetract(ConstraintStream stream,
-            List<? extends AbstractConstraintStream> childStreamList,
-            Function<Consumer<Tuple_>, AbstractInserter<Tuple_>> inserterConstructor,
-            BiFunction<Consumer<Tuple_>, Consumer<Tuple_>, AbstractUpdater<Tuple_>> updaterConstructor) {
-        Consumer<Tuple_> insert = getAggregatedInsert(childStreamList);
-        Consumer<Tuple_> update = getAggregatedUpdate(childStreamList);
-        Consumer<Tuple_> retract = getAggregatedRetract(childStreamList);
-        putInsertUpdateRetract(stream,
-                inserterConstructor.apply(insert),
-                updaterConstructor.apply(update, retract),
-                retract);
-    }
-
-    public <Tuple_ extends Tuple> Consumer<Tuple_> getAggregatedInsert(List<? extends ConstraintStream> streamList) {
-        return getAggregatedConsumer(streamList, insertMap);
+    public void addNode(AbstractNode node, ConstraintStream leftParent, ConstraintStream rightParent) {
+        addNode(node);
+        putInsertUpdateRetract(leftParent, TupleLifecycle.ofLeft((LeftTupleLifecycle<? extends Tuple>) node));
+        putInsertUpdateRetract(rightParent, TupleLifecycle.ofRight((RightTupleLifecycle<? extends Tuple>) node));
     }
```

Comment:
And this is the case where we add an extra indirection. Joins and ifExists can not implement TupleLifecycle, because they have different left and right entry points, so we need to do this.

With Gizmo, we could probably generate this away. But benchmarking does not show this as a problem.